room.go

```diff
@@ -1011,6 +1020,118 @@ func (r *Room) handleIncomingRpcRequest(callerIdentity, requestId, method, paylo
+	if _, ok := r.byteStreamHandlers.LoadOrStore(topic, handler); !ok {
+		return fmt.Errorf("byte stream handler already registered for topic: %s", topic)
+	}
+	return nil
+}
+
+func (r *Room) UnregisterByteStreamHandler(topic string) {
+	r.byteStreamHandlers.Delete(topic)
+}
+
+func (r *Room) handleStreamHeader(streamHeader *livekit.DataStream_Header, participantIdentity string) {
+	switch header := streamHeader.ContentHeader.(type) {
+	case *livekit.DataStream_Header_TextHeader:
+		streamHandlerCallback, ok := r.textStreamHandlers.Load(streamHeader.Topic)
+		if !ok {
+			r.log.Debugw("ignoring incoming text stream due to no handler for topic", "topic", streamHeader.Topic)
+			return
+		}
+
+		info := TextStreamInfo{
+			baseStreamInfo: &baseStreamInfo{
+				Id:         streamHeader.StreamId,
+				MimeType:   streamHeader.MimeType,
+				Size:       streamHeader.TotalLength,
+				Topic:      streamHeader.Topic,
+				Timestamp:  streamHeader.Timestamp,
+				Attributes: streamHeader.Attributes,
+			},
+		}
+
+		textStreamReader := NewTextStreamReader(info, streamHeader.TotalLength)
+		r.textStreamReaders.Store(streamHeader.StreamId, textStreamReader)
+		go streamHandlerCallback.(TextStreamHandler)(textStreamReader, participantIdentity)
+	case *livekit.DataStream_Header_ByteHeader:
+		streamHandlerCallback, ok := r.byteStreamHandlers.Load(streamHeader.Topic)
+		if !ok {
+			r.log.Debugw("ignoring incoming byte stream due to no handler for topic", "topic", streamHeader.Topic)
+			return
+		}
+
+		info := ByteStreamInfo{
+			baseStreamInfo: &baseStreamInfo{
+				Id:         streamHeader.StreamId,
+				MimeType:   streamHeader.MimeType,
+				Size:       streamHeader.TotalLength,
+				Topic:      streamHeader.Topic,
+				Timestamp:  streamHeader.Timestamp,
+				Attributes: streamHeader.Attributes,
+			},
+			Name: &header.ByteHeader.Name,
+		}
+
+		byteStreamReader := NewByteStreamReader(info, streamHeader.TotalLength)
+		r.byteStreamReaders.Store(streamHeader.StreamId, byteStreamReader)
+		go streamHandlerCallback.(ByteStreamHandler)(byteStreamReader, participantIdentity)
+	}
+}
+
+func (r *Room) handleStreamChunk(streamChunk *livekit.DataStream_Chunk) {
+	streamId := streamChunk.StreamId
+
+	byteStreamReader, ok := r.byteStreamReaders.Load(streamId)
+	if ok {
+		if len(streamChunk.Content) > 0 {
+			byteStreamReader.(*ByteStreamReader).enqueue(streamChunk)
+		}
+	}
+
+	textStreamReader, ok := r.textStreamReaders.Load(streamId)
+	if ok {
+		if len(streamChunk.Content) > 0 {
+			textStreamReader.(*TextStreamReader).enqueue(streamChunk)
+		}
+	}
+}
+
+func (r *Room) handleStreamTrailer(streamTrailer *livekit.DataStream_Trailer) {
+	streamId := streamTrailer.StreamId
+
+	byteStreamReader, ok := r.byteStreamReaders.Load(streamId)
+	if ok {
+		reader := byteStreamReader.(*ByteStreamReader)
+		for k, v := range streamTrailer.Attributes {
+			reader.Info.Attributes[k] = v
+		}
+		reader.close()
```

Comment:
The reader is closed immediately after setting the attributes. What is the purpose of it? Who uses those attributes?

Reply:
I'm not sure about this, I saw the JS SDK doing it. Maybe @lukasIO can help.

Reply:
I see that `ReadAll()` waits for close. So, I guess this is a signal for `ReadAll` to unblock and read?

Reply:
yeah that's for close yeah, I'm not sure what the trailer attributes are for

Reply:
JS uses the inherent mutability of objects to update attributes on the stream's info with additional attributes (or overrides) that are sent in the trailer. 
The idea was that e.g. for transcriptions when you're beginning a stream only the `startTime` is known. Only once the stream is finishing you'll have knowledge about the `endTime` so we allowed for updating attributes also as part of the trailer. 

In JS accessing `attributes` before and after a stream has closed can have different attributes set for that reason